### PR TITLE
Sentry Isolate Extension

### LINKED ISSUE #608: Provide functionality to make catching errors of other isolates easier
> This comes out of a discussion from https://github.com/getsentry/sentry-dart/issues/607

Currently it's not that easy to catch exceptions and errors in isolates spawned by the user.
It would be nice, if this can be made easier.

A couple of ideas:

## 1. Create a `SentrySendPort`
An isolate can be spawned by [`Isolate.spawn(entryPoint, "message");`](https://api.dart.dev/stable/2.14.3/dart-isolate/Isolate/spawn.html). It's possible to add a parameter for the `onError`-SendPort. By adding a `SentrySendPort` we can then do something like `Isolate.spawn(entryPoint, "message", onError: SentrySendPort());`.

## 2. Create extension method for isolates
Add an extension for isolates so that you can do `isolate.catchExceptionsWithSentry();` (must be called on an isolate instance). This can be achieved by using the `SentrySendPort` and [`addErrorListener(SendPort)`](https://api.dart.dev/stable/2.14.3/dart-isolate/Isolate/addErrorListener.html). 
Dart has no static extension methods, so something like `Isolate.spawnWithSentry(...)` is not possible.

### PR DESCRIPTION
**Pull request by denrase**
## :scroll: Description

Add extension & helper to handle isolate errors with Sentry.

Closes #608 

## :green_heart: How did you test it?

Moved existing tests.

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPii` is enabled
- [ ] I updated the docs if needed
- [x] All tests passing
- [x] No breaking changes

### CODE REVIEW

File: dart/lib/src/sentry_isolate.dart [hunk 1]
```diff
@@ -0,0 +1,95 @@
+import 'dart:async';
+import 'dart:isolate';
+
+import 'package:meta/meta.dart';
+
+import '../sentry.dart';
+
+/// Record isolate errors with the Sentry SDK.
+extension SentryIsolate on Isolate {
```
Comment (marandaneto) [inline dart/lib/src/sentry_isolate.dart]:
```suggestion
extension SentryIsolateExtension on Isolate {
```

File: dart/lib/src/sentry_isolate.dart [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+    final port = _createPort(hub, options);
+    addErrorListener(port.sendPort);
+    return port;
+  }
+
+  /// Pass the [receivePort] returned from [addSentryErrorListener] to remove
+  /// the sentry error listener.
+  void removeSentryErrorListener(RawReceivePort receivePort) {
+    receivePort.close();
```
Comment (marandaneto) [inline dart/lib/src/sentry_isolate.dart]: I believe we should not call `close` but` rather only remove the listener, the method is called `removeSentryErrorListener`

File: dart/lib/src/sentry_isolate.dart [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+extension SentryIsolate on Isolate {
+
+  /// Calls [addErrorListener] with an error listener from the Sentry SDK. Store
+  /// the returned [RawReceivePort] if you want to remove the Sentry listener
+  /// again.
+  ///
+  /// Since isolates run concurrently, it's possible for it to exit before the
+  /// error listener is established. To avoid this, start the isolate paused,
+  /// add the listener and then resume the isolate.
+  RawReceivePort addSentryErrorListener() {
```
Comment (marandaneto) [inline dart/lib/src/sentry_isolate.dart]: To make it testable, we could take the `hub` as an optional parameter and use `HubAdapter` by default if none provided.

File: dart/lib/src/sentry_isolate.dart (hunk 1 above)
Comment (marandaneto) [reply dart/lib/src/sentry_isolate.dart]: I'd rename the file as well

File: dart/lib/src/sentry_isolate.dart (hunk 2 above)
Comment (denrase) [reply dart/lib/src/sentry_isolate.dart]: We did do this in the integration, it's ok to change this behaviour?

Comment (marandaneto) [reply dart/lib/src/sentry_isolate.dart]: The integration should call `close` in the `receivePort`, but not the extension itself.
The integration is alive during app's lifecycle.
The extension can be used specifically for a few things.

**Comment by marandaneto** [comment]
@denrase are we adding extensions to `Isolate.spawn` on this PR as well?

**Comment by denrase** [comment]
@marandaneto I will try to come up with something. Since there are no static extensions, we might introduce some additional helper so users don't have to do the spawn paused, add observer, resume isolate dance boilerplate.

File: dart/lib/src/sentry_isolate.dart [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+  ///
+  /// Providing your own `onError` will not add the listener from Sentry SDK.
+  static Future<Isolate> spawn<T>(
+      void Function(T message) entryPoint, T message,
+      {bool paused = false,
+      bool errorsAreFatal = true,
+      SendPort? onExit,
+      SendPort? onError,
+      String? debugName,
+      Hub? hub}) async {
```
Comment (marandaneto) [inline dart/lib/src/sentry_isolate.dart]:
```suggestion
      @internal Hub? hub}) async {
```

File: dart/lib/src/sentry_isolate_extension.dart [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+/// Record isolate errors with the Sentry SDK.
+extension SentryIsolateExtension on Isolate {
+  /// Calls [addErrorListener] with an error listener from the Sentry SDK. Store
+  /// the returned [RawReceivePort] if you want to remove the Sentry listener
+  /// again.
+  ///
+  /// Since isolates run concurrently, it's possible for it to exit before the
+  /// error listener is established. To avoid this, start the isolate paused,
+  /// add the listener and then resume the isolate.
+  RawReceivePort addSentryErrorListener({Hub? hub}) {
```
Comment (marandaneto) [inline dart/lib/src/sentry_isolate_extension.dart]:
```suggestion
  RawReceivePort addSentryErrorListener({@internal Hub? hub}) {
```

File: dart/lib/src/sentry_isolate.dart [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+      message,
+      paused: paused,
+      errorsAreFatal: errorsAreFatal,
+      onExit: onExit,
+      onError: onError ?? createPort(hubToUse, options).sendPort,
+      debugName: debugName,
+    );
+  }
+
+  static RawReceivePort createPort(Hub hub, SentryOptions options) {
```
Comment (marandaneto) [inline dart/lib/src/sentry_isolate.dart]:
```suggestion
  @internal
  static RawReceivePort createPort(Hub hub, SentryOptions options) {
```

Comment (marandaneto) [inline dart/lib/src/sentry_isolate.dart]: You can access `hub.options` so you don't need to take `options` as well.

**Review by marandaneto — APPROVED** [review]
Left a few comments but overall pretty good.
@denrase thanks.

**Pull request merged**